### PR TITLE
Parse date range

### PR DESCRIPTION
**Pull request by laug**
## Description
**Linked issue**: #3596 #3906 #4002 

**Problem**
When selectsRange is false, users can input "03/04/2024" in text box and automatically have 4 March 2024 selected in the calendar.
However when selectsRange is true, if users input "03/04/2024 - 05/06/2024" in the text box, nothing happens.

**Changes**
With this PR, when selectsRange is true and the user inputs "03/04/2024 - 05/06/2024", the start date will be set to 4 March 2024 and the end date will be set to 6 May 2024.

## Contribution checklist
- [X] I have followed the [contributing guidelines](https://github.com/Hacker0x01/react-datepicker/blob/main/CONTRIBUTING.md).
- [X] I have added sufficient test coverage for my changes.
- [X] I have formatted my code with Prettier and checked for linting issues with ESLint for code readability.

### CODE REVIEW

File: src/date_utils.ts [hunk 1]
```diff
@@ -794,7 +794,7 @@ export function isDayDisabled(
         if (excludeDate instanceof Date) {
           return isSameDay(day, excludeDate);
         } else {
-          return isSameDay(day, excludeDate.date ?? new Date());
+          return isSameDay(day, excludeDate.date);
```
Comment (laug) [inline src/date_utils.ts:797]: When determining whether to disable a day due to it being in the list of excluded dates, it is not correct to check whether `day` is the same day as `new Date()` (today) as today is not necessarily in the list of excluded dates.
If `day` is non null and `excludeDate.date` is null, isSameDay will return false, which is correct.

File: src/index.tsx [hunk 2]
```diff
@@ -585,36 +586,77 @@ export default class DatePicker extends Component<
...
+      );
+      const startChanged = startDate?.getTime() !== startDateNew?.getTime();
+      const endChanged = endDate?.getTime() !== endDateNew?.getTime();
+
+      if (!startChanged && !endChanged) {
+        return;
+      }
+
+      if (startDateNew && isDayDisabled(startDateNew, this.props)) {
+        return;
```
Comment (martijnrusschen) [inline src/index.tsx:625]: Can you check if there's test coverage for this case?

File: src/index.tsx [hunk 3]
```diff
@@ -585,36 +586,77 @@ export default class DatePicker extends Component<
...
+        valueEnd ?? "",
+        dateFormat,
+        this.props.locale,
+        strictParsing,
+      );
+      const startChanged = startDate?.getTime() !== startDateNew?.getTime();
+      const endChanged = endDate?.getTime() !== endDateNew?.getTime();
+
+      if (!startChanged && !endChanged) {
+        return;
```
Comment (martijnrusschen) [inline src/index.tsx:621]: Can you check if there's test coverage for this case?

File: src/index.tsx [hunk 4]
```diff
@@ -585,36 +586,77 @@ export default class DatePicker extends Component<
...
+      }
+
+      if (startDateNew && isDayDisabled(startDateNew, this.props)) {
+        return;
+      }
+      if (endDateNew && isDayDisabled(endDateNew, this.props)) {
+        return;
```
Comment (martijnrusschen) [inline src/index.tsx:628]: Can you check if there's test coverage for this case?

Comment (laug) [reply src/index.tsx:628]: I've added coverage for the 3 early returns in commit 55f49a0, so it should be good now hopefully.

**Pull request merged**